File: Ballast.Core/src/Client/models/cubic-coordinates.ts
import { AxialCoordinates, IAxialCoordinates } from './axial-coordinates';
import { IOffsetCoordinates, OffsetCoordinates } from './offset-coordinates';

export interface ICubicCoordinates extends IAxialCoordinates {
    y: number;
}

export class CubicCoordinates implements ICubicCoordinates {

    public readonly x: number;
    public readonly y: number;
    public readonly z: number;

    private constructor(state: ICubicCoordinates) {
        if ((state.x + state.y + state.z) != 0) {
            throw new Error("Provided object coordinate(s) do not match constraint 'x + y + z = 0'");
        }
        this.x = state.x;
        this.y = state.y;
        this.z = state.z;
        if (this.x == -0)
            this.x = 0;
        if (this.y == -0)
            this.y = 0;
        if (this.z == -0)
            this.z = 0;
    }

    public static fromObject(object: ICubicCoordinates) {
        return new CubicCoordinates(object);
    }

    public static fromAxial(object: IAxialCoordinates) {
        let y = -1 * (object.x + object.z);
        return new CubicCoordinates({ x: object.x, y: y, z: object.z });
    }

    public static fromOffset(object: IOffsetCoordinates) {
        var x = object.col - (object.row - (object.row & 1)) / 2;
        var z = object.row;
        var y = -1 * (x + z);
        return new CubicCoordinates({ x: x, y: y, z: z });
    }

    public static fromOrderedTriple(orderedTriple: number[]) {
        if (orderedTriple.length < 3)
            throw new Error('Length of ordered pair must be 3 (or greater)');
        let x = orderedTriple[0];
        let y = orderedTriple[1];
        let z = orderedTriple[2];
        return CubicCoordinates.fromObject({ x: x, y: y, z: z });
    }

    public equals(object: ICubicCoordinates) {
        if (!object) {
            return false;
        }
        return (
            this.x == object.x &&
            this.y == object.y &&
            this.z == object.z
        );
    }

    public equalsAxial(object: IAxialCoordinates) {
        return CubicCoordinates
            .fromAxial(object)
            .equals(this);
    }

    public equalsOffset(object: IOffsetCoordinates) {
        return CubicCoordinates
            .fromOffset(object)
            .equals(this);
    }

    public toAxial() {
        return AxialCoordinates.fromCubic(this);
    }

    public toOffset() {
        return OffsetCoordinates.fromCubic(this);
    }

    public toOrderedTriple() {
        return [this.x, this.y, this.z];
    }

}

File: Ballast.Core/src/Server/Models/CubicCoordinates.cs
using System;

namespace Ballast.Core.Models
{
    public class CubicCoordinates : ICubicCoordinates
    {
        public int X { get; private set; }
        public int Y { get; private set; }
        public int Z { get; private set; }

        private CubicCoordinates(int x, int y, int z)
        {
            if ((x + y + z) != 0)
                throw new Exception("Provided object coordinate(s) do not match constraint 'x + y + z = 0'");
            X = x;
            Y = y;
            Z = z;
        }

        private CubicCoordinates(ICubicCoordinates state) : this(
            x: state.X, 
            y: state.Y, 
            z: state.Z
        ) { }

        public static CubicCoordinates FromProperties(int x, int y, int z) => new CubicCoordinates(
            x: x,
            y: y,
            z: z
        );

        public static CubicCoordinates FromObject(ICubicCoordinates state) => new CubicCoordinates(state);

        public static CubicCoordinates FromAxial(IAxialCoordinates state)
        {
            var y = -1 * (state.X + state.Z);
            return CubicCoordinates.FromProperties(
                x: state.X,
                y: y,
                z: state.Z
            );
        }

        public static CubicCoordinates FromOffset(IOffsetCoordinates state)
        {
            var x = state.Col - (state.Row - (state.Row & 1)) / 2;
            var z = state.Row;
            var y = -1 * (x + z);
            return CubicCoordinates.FromProperties(x: x, y: y, z: z);
        }

        public static CubicCoordinates FromOrderedTriple(int[] orderedTriple)
        {
            if (orderedTriple.Length < 3)
                throw new Exception("Length of ordered pair must be 3 (or greater)");
            var x = orderedTriple[0];
            var y = orderedTriple[1];
            var z = orderedTriple[2];
            return CubicCoordinates.FromProperties(x: x, y: y, z: z);
        }

        public bool Equals(ICubicCoordinates value)
        {
            if (value == null)
                return false;
            return (
                X == value.X &&
                Y == value.Y &&
                Z == value.Z
            );
        }

        public bool EqualsAxial(IAxialCoordinates value) =>
            CubicCoordinates
                .FromAxial(value)
                .Equals(this);

        public bool EqualsOffset(IOffsetCoordinates value) =>
            CubicCoordinates
                .FromOffset(value)
                .Equals(this);

        public AxialCoordinates ToAxial()
        {
            return AxialCoordinates.FromCubic(this);
        }

        public OffsetCoordinates ToOffset()
        {
            return OffsetCoordinates.FromCubic(this);
        }

        public int[] ToOrderedTriple()
        {
            return new int[] { X, Y, Z };
        }

    }

}